log4j-core/src/main/java/org/apache/logging/log4j/core/pattern/NameAbbreviator.java

```diff
@@ -285,40 +285,40 @@ public PatternAbbreviatorFragment(
         /**
          * Abbreviate element of name.
          *
-         * @param buf      buffer to receive element.
-         * @param startPos starting index of name element.
-         * @return starting index of next element.
+         * @param input      input string which is being written to the output {@code buf}.
+         * @param inputIndex starting index of name element in the {@code input} string.
+         * @param buf        buffer to receive element.
+         * @return starting  index of next element.
          */
-        public int abbreviate(final StringBuilder buf, final int startPos) {
-            final int start = (startPos < 0) ? 0 : startPos;
-            final int max = buf.length();
-            int nextDot = -1;
-            for (int i = start; i < max; i++) {
-                if (buf.charAt(i) == '.') {
-                    nextDot = i;
-                    break;
-                }
+        int abbreviate(final String input, final int inputIndex, final StringBuilder buf) {
+            // ckozak: indexOf with a string is intentional. indexOf(String, int) appears to
+            // have optimizations that don't apply to indexOf(char, int)
+            // which result in a >10% performance improvement in some
+            // NamePatternConverterBenchmark cases. This should be re-evaluated with
+            // future java releases.
+            int nextDot = input.indexOf(".", inputIndex);
```

Comment:
This performance change was _wild_:

```diff
- int nextDot = input.indexOf('.', inputIndex);
+ int nextDot = input.indexOf(".", inputIndex);
```

@garydgregory Any chance you have ideas that might explain this? It threw me for a loop because the character-based indexOf underperforms compared to the old implementation, while the string-based indexOf has an advantage and much lower variance.

Reply:
I've not looked at this but the scary part to me is that the performance might vary from JRE vendor to JRE vendor and from version to version.

Reply:
Agreed. I verified on linux x86_64 with openjdk 11 and 15, but it's likely performance is different elsewhere (although not clear in which direction it differs).

Reply:
Likely due to difference with intrinsic for `indexOf` with `String`, while the `char` version is not intrinsified until OpenJDK 16+, see https://bugs.openjdk.java.net/browse/JDK-8173585 , https://github.com/openjdk/jdk/commit/f71e8a61 and https://mail.openjdk.java.net/pipermail/jdk9-dev/2017-January/005539.html

Reply:
Great find! I really appreciate your insight, @schlosna. I'll a/b test on jdk16 to see how performance is impacted. If the character search is much better, it might be worthwhile to use that as it should stand the test of time, especially with jdk17 right around the corner.